### PR TITLE
KAFKA-18026: KIP-1112, migrate foreign-key joins to use ProcesserSupplier#stores 

### PR DESCRIPTION
**Pull request by ableegoldman**
Convert FKJ processors to implementing the #stores method

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:1281]: Just curious is the final goal to eventually remove `StatefulProcessorNode` and only use `ProcessorGraphNode`?

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:2214]: What the comment `// 6` is about? I think this test case still only have 5 store?

**Review by guozhangwang — APPROVED** [review]
One nit question about unit test, please feel free to merge afterwards.

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:2214]: copy-paste error 🙃 thanks

And yeah, it should have 5 stores

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:1281]: good question -- basically yes and no. Almost everything has/will be converted from StatefulProcesorNode to ProcessorGraphNode, but there are a few exceptions where we still need the StatefulProcessorNode to connect a processor and state store. For example any processors that access upstream state via a value getter, or a custom ProcessorSupplier passed into the `.process`/`.processValues` operator that doesn't implement `#stores` and manually adds stores by calling `StreamsBuilder#addStateStore` instead.

I actually have the PR for this ready but it was done on top of this so I was waiting for this one to be merged before calling for review. But it's ready now if you want to take a look (though note that it still can't be merged until Almog's TableSuppressNode PR is merged). See https://github.com/apache/kafka/pull/18195

**Comment by ableegoldman** [comment]
test failure is unrelated, merging to trunk

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:2214]: ill address this in my next PR so i doin't have to run the tests again before merging

**Pull request merged**
